src/hooks/analytics.jsx

```diff
@@ -0,0 +1,99 @@
+import {
+  createRef, useCallback, useContext, useMemo,
+} from 'react';
+
+import { ParagonContext } from '../ParagonProvider';
+
+export const ANALYTICS_TIMEOUT_MS = 300;
+
+/**
+ * Returns the ``analytics`` key stored in ParagonProvider's context
+ */
+export const useAnalytics = () => {
+  const { analytics } = useContext(ParagonContext) || {};
+  return analytics;
+};
+
+/**
+ *
+ * @param {object} [event] Object containing the analytic event name, properties, and dispatchers.
+ * @param {function} [onClick] Custom event handler to call when processing the onClick event.
+ * @param {object} [forwardedRef] An optional React forwarded ref.
+ *
+ * @returns {array} Array containing an onClick event handler to pass to an element that will dispatch
+ *  an analytics event, and an element ref.
+ */
+export const useHandleLogClick = ({
+  event, onClick, forwardedRef,
+}) => {
+  const ref = useMemo(
+    () => forwardedRef || createRef(),
+    [forwardedRef],
+  );
+  const analytics = useAnalytics();
+  const {
+    name: eventName,
+    properties: eventProperties,
+    dispatchers: eventDispatchers,
+  } = event || {};
+
+  /**
+   * Iterates through all analytic event dispatch functions passed in, passing the event name and
+   * properties to each.
+   */
+  const sendEventToDispatchers = useCallback(
+    () => {
+      if (!eventDispatchers) {
+        return;
+      }
+      eventDispatchers.filter(dispatcher => typeof dispatcher === 'function').forEach((dispatch) => {
+        dispatch(eventName, eventProperties);
+      });
+    },
+    [eventName, eventProperties, eventDispatchers],
+  );
+
+  /**
+   * Sends an analytic event to ``analytics.sendTrackEvent`` if it exists as well as sending the
+   * same event to the provided custom event dispatchers.
+   */
+  const dispatchAnalyticEvents = () => {
+    if (analytics?.sendTrackEvent) {
+      analytics.sendTrackEvent(eventName, eventProperties);
+    }
+    sendEventToDispatchers();
+  };
+
+  /**
+   * An onClick event handler that appropriately configures an element to dispatch an analytic event.
+   * If the element has an `href` (i.e., hyperlink), delay execution of the location redirect to give
+   * enough time for the analytic event to be dispatched. Otherwise, dispatch the analytic event immediately.
+   *
+   * @param {event} e The event associated with onClick.
+   */
+  const handleLogClick = (e) => {
+    if (event) {
+      // handle clicked anchor links that open in same tab by adding a slight delay
+      // to the page redirect to allow enough time for an analytics event to
+      // be dispatched.
+      if (ref?.current?.href && ref?.current?.target !== '_blank') {
+        e.preventDefault();
+        dispatchAnalyticEvents();
+        setTimeout(() => {
+          global.location.href = ref.current.href;
+        }, [ANALYTICS_TIMEOUT_MS]);
```

Comment:
Hm, this seems brittle... I'm pretty sure the analytics network request will get canceled if this fires before it's complete, which for slow connections might happen frequently at 300ms.  

And unfortunately for fast internet users, 300ms is sorta introducing a lot of unnecessary lag into the system. :/

Reply:
This is based on how Segment implements their `trackLink` method, [which also introduces a 300ms delay](https://github.com/segmentio/analytics.js-core/blob/4f3035c5e959ea92d5c31a152211654242b23e00/lib/analytics.ts#L516). This delay gives enough time for the network request to resolve without being noticeably perceived by users.

If a network request doesn't resolve before the delay, we are prioritizing the UX over ensuring the event is logged to not introduce any more of a delay than necessary.

Reply:
Also, FWIW, I did try to use `trackLink` here but it wasn't usable as is because it creates duplicate `onClick` event handlers on the provided element any time a prop changes with no clean way to remove the duplicate event handlers on re-renders.